File: sdks/thunkmetrc-py/wrapper/src/thunkmetrc/wrapper/wrapper.py
from typing import Optional
from thunkmetrc.client import MetrcClient
from .ratelimiter import MetrcRateLimiter, RateLimiterConfig
from .services import *

class MetrcWrapper:
    def __init__(self, client: MetrcClient, config: Optional[RateLimiterConfig] = None):
        self.client = client
        self.limiter = MetrcRateLimiter(config)
        self.additives_templates = AdditivesTemplatesService(self.client, self.limiter)
        self.caregivers_status = CaregiversStatusService(self.client, self.limiter)
        self.employees = EmployeesService(self.client, self.limiter)
        self.facilities = FacilitiesService(self.client, self.limiter)
        self.harvests = HarvestsService(self.client, self.limiter)
        self.items = ItemsService(self.client, self.limiter)
        self.lab_tests = LabTestsService(self.client, self.limiter)
        self.locations = LocationsService(self.client, self.limiter)
        self.packages = PackagesService(self.client, self.limiter)
        self.patient_check_ins = PatientCheckInsService(self.client, self.limiter)
        self.patients = PatientsService(self.client, self.limiter)
        self.patients_status = PatientsStatusService(self.client, self.limiter)
        self.plant_batches = PlantBatchesService(self.client, self.limiter)
        self.plants = PlantsService(self.client, self.limiter)
        self.processing_job = ProcessingJobService(self.client, self.limiter)
        self.retail_id = RetailIdService(self.client, self.limiter)
        self.sales = SalesService(self.client, self.limiter)
        self.sandbox = SandboxService(self.client, self.limiter)
        self.strains = StrainsService(self.client, self.limiter)
        self.sublocations = SublocationsService(self.client, self.limiter)
        self.tags = TagsService(self.client, self.limiter)
        self.transfers = TransfersService(self.client, self.limiter)
        self.transporters = TransportersService(self.client, self.limiter)
        self.units_of_measure = UnitsOfMeasureService(self.client, self.limiter)
        self.waste_methods = WasteMethodsService(self.client, self.limiter)
        self.webhooks = WebhooksService(self.client, self.limiter)
    @property
    def AdditivesTemplates(self) -> AdditivesTemplatesService:
        return self.additives_templates
    @property
    def CaregiversStatus(self) -> CaregiversStatusService:
        return self.caregivers_status
    @property
    def Employees(self) -> EmployeesService:
        return self.employees
    @property
    def Facilities(self) -> FacilitiesService:
        return self.facilities
    @property
    def Harvests(self) -> HarvestsService:
        return self.harvests
    @property
    def Items(self) -> ItemsService:
        return self.items
    @property
    def LabTests(self) -> LabTestsService:
        return self.lab_tests
    @property
    def Locations(self) -> LocationsService:
        return self.locations
    @property
    def Packages(self) -> PackagesService:
        return self.packages
    @property
    def PatientCheckIns(self) -> PatientCheckInsService:
        return self.patient_check_ins
    @property
    def Patients(self) -> PatientsService:
        return self.patients
    @property
    def PatientsStatus(self) -> PatientsStatusService:
        return self.patients_status
    @property
    def PlantBatches(self) -> PlantBatchesService:
        return self.plant_batches
    @property
    def Plants(self) -> PlantsService:
        return self.plants
    @property
    def ProcessingJob(self) -> ProcessingJobService:
        return self.processing_job
    @property
    def RetailId(self) -> RetailIdService:
        return self.retail_id
    @property
    def Sales(self) -> SalesService:
        return self.sales
    @property
    def Sandbox(self) -> SandboxService:
        return self.sandbox
    @property
    def Strains(self) -> StrainsService:
        return self.strains
    @property
    def Sublocations(self) -> SublocationsService:
        return self.sublocations
    @property
    def Tags(self) -> TagsService:
        return self.tags
    @property
    def Transfers(self) -> TransfersService:
        return self.transfers
    @property
    def Transporters(self) -> TransportersService:
        return self.transporters
    @property
    def UnitsOfMeasure(self) -> UnitsOfMeasureService:
        return self.units_of_measure
    @property
    def WasteMethods(self) -> WasteMethodsService:
        return self.waste_methods
    @property
    def Webhooks(self) -> WebhooksService:
        return self.webhooks

File: sdks/thunkmetrc-go/wrapper/wrapper.go
package wrapper

import (
	"log/slog"

	"github.com/thunkier/thunkmetrc/sdks/thunkmetrc-go/client"
	"github.com/thunkier/thunkmetrc/sdks/thunkmetrc-go/wrapper/services"
)

type MetrcWrapper struct {
	Client      *client.MetrcClient
	RateLimiter services.RateLimiter
	AdditivesTemplates *services.AdditivesTemplatesService
	CaregiversStatus *services.CaregiversStatusService
	Employees *services.EmployeesService
	Facilities *services.FacilitiesService
	Harvests *services.HarvestsService
	Items *services.ItemsService
	LabTests *services.LabTestsService
	Locations *services.LocationsService
	Packages *services.PackagesService
	PatientCheckIns *services.PatientCheckInsService
	Patients *services.PatientsService
	PatientsStatus *services.PatientsStatusService
	PlantBatches *services.PlantBatchesService
	Plants *services.PlantsService
	ProcessingJob *services.ProcessingJobService
	RetailId *services.RetailIdService
	Sales *services.SalesService
	Sandbox *services.SandboxService
	Strains *services.StrainsService
	Sublocations *services.SublocationsService
	Tags *services.TagsService
	Transfers *services.TransfersService
	Transporters *services.TransportersService
	UnitsOfMeasure *services.UnitsOfMeasureService
	WasteMethods *services.WasteMethodsService
	Webhooks *services.WebhooksService
}

func New(c *client.MetrcClient) *MetrcWrapper {
	rl := services.NewRateLimiter(services.DefaultConfig(), c.Logger)
	return NewWithLimiter(c, rl)
}

func NewMetrcWrapper(baseUrl string, accessKey string, logger *slog.Logger) *MetrcWrapper {
	c := client.New(baseUrl, accessKey, "")
	rl := services.NewRateLimiter(services.DefaultConfig(), logger)

	return NewWithLimiter(c, rl)
}

func NewWithLimiter(client *client.MetrcClient, rateLimiter services.RateLimiter) *MetrcWrapper {
	return &MetrcWrapper{
		Client:      client,
		RateLimiter: rateLimiter,
			AdditivesTemplates: &services.AdditivesTemplatesService{Client: client, RateLimiter: rateLimiter},
			CaregiversStatus: &services.CaregiversStatusService{Client: client, RateLimiter: rateLimiter},
			Employees: &services.EmployeesService{Client: client, RateLimiter: rateLimiter},
			Facilities: &services.FacilitiesService{Client: client, RateLimiter: rateLimiter},
			Harvests: &services.HarvestsService{Client: client, RateLimiter: rateLimiter},
			Items: &services.ItemsService{Client: client, RateLimiter: rateLimiter},
			LabTests: &services.LabTestsService{Client: client, RateLimiter: rateLimiter},
			Locations: &services.LocationsService{Client: client, RateLimiter: rateLimiter},
			Packages: &services.PackagesService{Client: client, RateLimiter: rateLimiter},
			PatientCheckIns: &services.PatientCheckInsService{Client: client, RateLimiter: rateLimiter},
			Patients: &services.PatientsService{Client: client, RateLimiter: rateLimiter},
			PatientsStatus: &services.PatientsStatusService{Client: client, RateLimiter: rateLimiter},
			PlantBatches: &services.PlantBatchesService{Client: client, RateLimiter: rateLimiter},
			Plants: &services.PlantsService{Client: client, RateLimiter: rateLimiter},
			ProcessingJob: &services.ProcessingJobService{Client: client, RateLimiter: rateLimiter},
			RetailId: &services.RetailIdService{Client: client, RateLimiter: rateLimiter},
			Sales: &services.SalesService{Client: client, RateLimiter: rateLimiter},
			Sandbox: &services.SandboxService{Client: client, RateLimiter: rateLimiter},
			Strains: &services.StrainsService{Client: client, RateLimiter: rateLimiter},
			Sublocations: &services.SublocationsService{Client: client, RateLimiter: rateLimiter},
			Tags: &services.TagsService{Client: client, RateLimiter: rateLimiter},
			Transfers: &services.TransfersService{Client: client, RateLimiter: rateLimiter},
			Transporters: &services.TransportersService{Client: client, RateLimiter: rateLimiter},
			UnitsOfMeasure: &services.UnitsOfMeasureService{Client: client, RateLimiter: rateLimiter},
			WasteMethods: &services.WasteMethodsService{Client: client, RateLimiter: rateLimiter},
			Webhooks: &services.WebhooksService{Client: client, RateLimiter: rateLimiter},
	}
}



